### ISSUE TITLE
wintun adapter error after "renew session"

### ISSUE THREAD

**Issue by efef** [Amebis/eduVPN#166]
While testing with vpn.spoor.nu, after using the "renew session" button I am getting this wintun fatal error notification:

![Screenshot 2021-06-16 at 16 05 16](https://user-images.githubusercontent.com/3393474/122540714-3072d280-d029-11eb-85e9-035292fbe0e2.png)

It is reproducible

**Comment by efef** [comment]
it seems the issue is only reproducible in VMware win10 VM which I downloaded from https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/
Am running win10 20H2 build 19042.1052
On a physical pc, with exact same build version, I don't experience any issue

**Comment by efef** [comment]
in windows event viewer I read:  openvpnserv$eduVPN error: An attempt was made to perform an initialization operation when initialization has already been completed. (0x4df) 
   Could not register ring buffers

**Comment by rozmansi** [comment]
I am able to reproduce this bug. Resolving...

**Comment by rozmansi** [comment]
Found the error and fixed it.

Wintun has an IOCTL to register rings. Rings are automatically unregistered when client closes all file handles to Wintun. OpenVPN has two handles to Wintun: one in openvpn.exe and one in the interactive service. My OpenVPN ⇿ Wintun interaction closed old handle 1/2 from the first connection. On reconnection, it opened new handle 1/2, then closed old handle 2/2, then opened new handle 2/2. Finally, when it tried to register Wintun rings, Wintun said, they're still registered.

If this sounds confusing, simply put: there was no time window when OpenVPN would close both handles to Wintun to allow Wintun to unregister rings.

I had to add extra iservice function, so openvpn.exe can ask iservice to close its Wintun handle too when appropriate.
